drivers/overlay/overlay.go

```diff
@@ -2020,11 +2021,27 @@ func (d *Driver) getWhiteoutFormat() archive.WhiteoutFormat {
 }
 
 type overlayFileGetter struct {
-	diffDirs []string
+	diffDirs        []string
+	composefsMounts map[string]*os.File // map from diff dir to the directory with the composefs blob mounted
 }
 
 func (g *overlayFileGetter) Get(path string) (io.ReadCloser, error) {
+	buf := make([]byte, unix.PathMax)
 	for _, d := range g.diffDirs {
+		if f, found := g.composefsMounts[d]; found {
+			// there is no *at equivalent for getxattr, but it can be emulated by opening the file under /proc/self/fd/$FD/$PATH
+			len, err := unix.Getxattr(fmt.Sprintf("/proc/self/fd/%d/%s", int(f.Fd()), path), "trusted.overlay.redirect", buf)
+			if err != nil {
+				continue
+			}
+
+			// the xattr value is the path to the file in the composefs layer diff directory
+			f, err := os.Open(filepath.Join(d, string(buf[:len])))
+			if err == nil {
```

Comment:
I wonder if we should really fall through here if we hit an error?  Under what scenario would that happen, and it would be the right thing to just do the non-composefs overlay path?

Reply:
yes that is a mistake, thanks for catching it.  There is no point in the fall through when it is a composefs layer.

I've changed it and pushed a new version